overlord/patch/patch.go

```diff
@@ -0,0 +1,107 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patch
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+// Level is the current implemented patch level of the state format and content.
+var Level = 1
+
+// patches maps from patch level L to the function that moves from L-1 to L.
+var patches = make(map[int]func(s *state.State) error)
+
+// Init initializes an empty state to the current implemented patch level.
+func Init(s *state.State) {
+	s.Lock()
+	defer s.Unlock()
+	if s.Get("patch-level", new(int)) != state.ErrNoState {
+		panic("internal error: expected empty state, attempting to override patch-level without actual patching")
+	}
+	s.Set("patch-level", Level)
+}
+
+// Apply applies any necessary patches to update the provided state to
+// conventions required by the current patch level of the system.
+func Apply(s *state.State) error {
+	var stateLevel int
+	s.Lock()
+	err := s.Get("patch-level", &stateLevel)
+	s.Unlock()
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if stateLevel == Level {
+		// already at right level, nothing to do
+		return nil
+	}
+	if stateLevel > Level {
+		return fmt.Errorf("cannot downgrade: snapd is too old for the current system state (patch level %d)", stateLevel)
```

Comment:
This is something that will make os-level rollbacks interessting :/ Not related to this branch just as an observation.